src/shr_transport.h4

```diff
@@ -175,23 +175,36 @@ shmem_shr_transport_put(shmem_ctx_t ctx, void *target, const void *source,
 static inline void
 shmem_shr_transport_put_signal(shmem_ctx_t ctx, void *target,
                                const void *source, size_t len,
-                               uint64_t *sig_addr, uint64_t signal, int pe)
+                               uint64_t *sig_addr, uint64_t signal, int sig_op, int pe)
 {
 #if USE_MEMCPY
     memcpy(target, source, len);
-    *sig_addr = signal;
+    if (sig_op == SHMEM_SIGNAL_ADD) *sig_addr += signal;
+    else *sig_addr = signal;
 #elif USE_XPMEM
     shmem_transport_xpmem_put(target, source, len, pe,
                               shmem_internal_get_shr_rank(pe));
     shmem_internal_membar_acq_rel(); /* Memory fence to ensure target PE observes
                                         stores in the correct order */
+    uint64_t old_signal_val = 0;
+    if (sig_op == SHMEM_SIGNAL_ADD) {
+        shmem_transport_xpmem_get(&old_signal_val, sig_addr, sizeof(uint64_t), pe,
+                                  shmem_internal_get_shr_rank(pe));
+        signal += old_signal_val;
+    }
     shmem_transport_xpmem_put(sig_addr, &signal, sizeof(uint64_t), pe,
                               shmem_internal_get_shr_rank(pe));
 #elif USE_CMA
     shmem_transport_cma_put(target, source, len, pe,
                             shmem_internal_get_shr_rank(pe));
     shmem_internal_membar_acq_rel(); /* Memory fence to ensure target PE observes
                                         stores in the correct order */
+    uint64_t old_signal_val = 0;
+    if (sig_op == SHMEM_SIGNAL_ADD) {
+        shmem_transport_cma_get(&old_signal_val, sig_addr, sizeof(uint64_t), pe,
+                                shmem_internal_get_shr_rank(pe));
+        signal += old_signal_val;
+    }
```

Comment:
This needs to be an atomic. CMA does not support atomics; use `shmem_transport_atomic`.